// electron-builder/out/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _builderUtil;

function _load_builderUtil() {
  return _builderUtil = require("builder-util");
}

Object.defineProperty(exports, "getArchSuffix", {
  enumerable: true,
  get: function () {
    return (_builderUtil || _load_builderUtil()).getArchSuffix;
  }
});
Object.defineProperty(exports, "Arch", {
  enumerable: true,
  get: function () {
    return (_builderUtil || _load_builderUtil()).Arch;
  }
});
Object.defineProperty(exports, "archFromString", {
  enumerable: true,
  get: function () {
    return (_builderUtil || _load_builderUtil()).archFromString;
  }
});

var _builder;

function _load_builder() {
  return _builder = require("./builder");
}

Object.defineProperty(exports, "build", {
  enumerable: true,
  get: function () {
    return (_builder || _load_builder()).build;
  }
});
Object.defineProperty(exports, "createTargets", {
  enumerable: true,
  get: function () {
    return (_builder || _load_builder()).createTargets;
  }
});

var _electronBuilderLib;

function _load_electronBuilderLib() {
  return _electronBuilderLib = require("electron-builder-lib");
}

Object.defineProperty(exports, "Platform", {
  enumerable: true,
  get: function () {
    return (_electronBuilderLib || _load_electronBuilderLib()).Platform;
  }
});
Object.defineProperty(exports, "Target", {
  enumerable: true,
  get: function () {
    return (_electronBuilderLib || _load_electronBuilderLib()).Target;
  }
});
Object.defineProperty(exports, "DIR_TARGET", {
  enumerable: true,
  get: function () {
    return (_electronBuilderLib || _load_electronBuilderLib()).DIR_TARGET;
  }
});
Object.defineProperty(exports, "DEFAULT_TARGET", {
  enumerable: true,
  get: function () {
    return (_electronBuilderLib || _load_electronBuilderLib()).DEFAULT_TARGET;
  }
});
Object.defineProperty(exports, "Packager", {
  enumerable: true,
  get: function () {
    return (_electronBuilderLib || _load_electronBuilderLib()).Packager;
  }
});
Object.defineProperty(exports, "AppInfo", {
  enumerable: true,
  get: function () {
    return (_electronBuilderLib || _load_electronBuilderLib()).AppInfo;
  }
});
Object.defineProperty(exports, "PublishManager", {
  enumerable: true,
  get: function () {
    return (_electronBuilderLib || _load_electronBuilderLib()).PublishManager;
  }
});

var _forgeMaker;

function _load_forgeMaker() {
  return _forgeMaker = require("./forge/forge-maker");
}

Object.defineProperty(exports, "buildForge", {
  enumerable: true,
  get: function () {
    return (_forgeMaker || _load_forgeMaker()).buildForge;
  }
});

var _builderUtilRuntime;

function _load_builderUtilRuntime() {
  return _builderUtilRuntime = require("builder-util-runtime");
}

Object.defineProperty(exports, "CancellationToken", {
  enumerable: true,
  get: function () {
    return (_builderUtilRuntime || _load_builderUtilRuntime()).CancellationToken;
  }
});
