// electron-publish/out/publisher.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.HttpPublisher = exports.Publisher = undefined;

var _bluebirdLst;

function _load_bluebirdLst() {
    return _bluebirdLst = require("bluebird-lst");
}

exports.getCiTag = getCiTag;

var _chalk;

function _load_chalk() {
    return _chalk = require("chalk");
}

var _electronBuilderHttp;

function _load_electronBuilderHttp() {
    return _electronBuilderHttp = require("electron-builder-http");
}

var _builderUtil;

function _load_builderUtil() {
    return _builderUtil = require("builder-util");
}

var _fsExtraP;

function _load_fsExtraP() {
    return _fsExtraP = require("fs-extra-p");
}

var _path = require("path");

const progressBarOptions = {
    incomplete: " ",
    width: 20
};
class Publisher {
    constructor(context) {
        this.context = context;
    }
    createProgressBar(fileName, fileStat) {
        if (this.context.progress == null) {
            (0, (_builderUtil || _load_builderUtil()).log)(`Uploading ${fileName} to ${this.providerName}`);
            return null;
        } else {
            return this.context.progress.createBar(`[:bar] :percent :etas | ${(0, (_chalk || _load_chalk()).green)(fileName)} to ${this.providerName}`, Object.assign({ total: fileStat.size }, progressBarOptions));
        }
    }
    createReadStreamAndProgressBar(file, fileStat, progressBar, reject) {
        const fileInputStream = (0, (_fsExtraP || _load_fsExtraP()).createReadStream)(file);
        fileInputStream.on("error", reject);
        if (progressBar == null) {
            return fileInputStream;
        } else {
            const progressStream = new (_electronBuilderHttp || _load_electronBuilderHttp()).ProgressCallbackTransform(fileStat.size, this.context.cancellationToken, it => progressBar.tick(it.delta));
            progressStream.on("error", reject);
            return fileInputStream.pipe(progressStream);
        }
    }
}
exports.Publisher = Publisher;
class HttpPublisher extends Publisher {
    constructor(context) {
        let useSafeArtifactName = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;

        super(context);
        this.context = context;
        this.useSafeArtifactName = useSafeArtifactName;
    }
    upload(file, arch, safeArtifactName) {
        var _this = this;

        return (0, (_bluebirdLst || _load_bluebirdLst()).coroutine)(function* () {
            const fileName = (_this.useSafeArtifactName ? safeArtifactName : null) || (0, _path.basename)(file);
            const fileStat = yield (0, (_fsExtraP || _load_fsExtraP()).stat)(file);
            const progressBar = _this.createProgressBar(fileName, fileStat);
            yield _this.doUpload(fileName, arch, fileStat.size, function (request, reject) {
                if (progressBar != null) {
                    // reset (because can be called several times (several attempts)
                    progressBar.update(0);
                }
                return _this.createReadStreamAndProgressBar(file, fileStat, progressBar, reject).pipe(request);
            }, file);
        })();
    }
    uploadData(data, arch, fileName) {
        if (data == null || fileName == null) {
            throw new Error("data or fileName is null");
        }
        return this.doUpload(fileName, arch, data.length, it => it.end(data));
    }
}
exports.HttpPublisher = HttpPublisher;
function getCiTag() {
    const tag = process.env.TRAVIS_TAG || process.env.APPVEYOR_REPO_TAG_NAME || process.env.CIRCLE_TAG || process.env.BITRISE_GIT_TAG || process.env.CI_BUILD_TAG;
    return tag != null && tag.length > 0 ? tag : null;
}
