// electron-builder-lib/out/targets/pkg.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.prepareProductBuildArgs = prepareProductBuildArgs;
exports.PkgTarget = void 0;

function _builderUtil() {
  const data = require("builder-util");

  _builderUtil = function () {
    return data;
  };

  return data;
}

function _fs() {
  const data = require("builder-util/out/fs");

  _fs = function () {
    return data;
  };

  return data;
}

function _license() {
  const data = require("builder-util/out/license");

  _license = function () {
    return data;
  };

  return data;
}

function _fsExtraP() {
  const data = require("fs-extra-p");

  _fsExtraP = function () {
    return data;
  };

  return data;
}

var path = _interopRequireWildcard(require("path"));

function _codeSign() {
  const data = require("../codeSign");

  _codeSign = function () {
    return data;
  };

  return data;
}

function _core() {
  const data = require("../core");

  _core = function () {
    return data;
  };

  return data;
}

function _mac() {
  const data = require("../packager/mac");

  _mac = function () {
    return data;
  };

  return data;
}

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

const certType = "Developer ID Installer"; // http://www.shanekirk.com/2013/10/creating-flat-packages-in-osx/
// to use --scripts, we must build .app bundle separately using pkgbuild
// productbuild --scripts doesn't work (because scripts in this case not added to our package)
// https://github.com/electron-userland/electron-osx-sign/issues/96#issuecomment-274986942

class PkgTarget extends _core().Target {
  constructor(packager, outDir) {
    super("pkg");
    this.packager = packager;
    this.outDir = outDir;
    this.options = Object.assign({
      allowAnywhere: true,
      allowCurrentUserHome: true,
      allowRootDirectory: true
    }, this.packager.config.pkg);
  }

  async build(appPath, arch) {
    const packager = this.packager;
    const options = this.options;
    const appInfo = packager.appInfo;
    const artifactName = packager.expandArtifactNamePattern(options, "pkg");
    const artifactPath = path.join(this.outDir, artifactName);
    this.logBuilding("pkg", artifactPath, arch);
    const keychainName = (await packager.codeSigningInfo.value).keychainName;
    const appOutDir = this.outDir; // https://developer.apple.com/library/content/documentation/DeveloperTools/Reference/DistributionDefinitionRef/Chapters/Distribution_XML_Ref.html

    const distInfoFile = path.join(appOutDir, "distribution.xml");
    const innerPackageFile = path.join(appOutDir, `${(0, _mac().filterCFBundleIdentifier)(appInfo.id)}.pkg`);
    const identity = (await Promise.all([(0, _codeSign().findIdentity)(certType, options.identity || packager.platformSpecificBuildOptions.identity, keychainName), this.customizeDistributionConfiguration(distInfoFile, appPath), this.buildComponentPackage(appPath, innerPackageFile)]))[0];

    if (identity == null && packager.forceCodeSigning) {
      throw new Error(`Cannot find valid "${certType}" to sign standalone installer, please see https://electron.build/code-signing`);
    }

    const args = prepareProductBuildArgs(identity, keychainName);
    args.push("--distribution", distInfoFile);
    args.push(artifactPath);
    (0, _builderUtil().use)(options.productbuild, it => args.push(...it));
    await (0, _builderUtil().exec)("productbuild", args, {
      cwd: appOutDir
    });
    await Promise.all([(0, _fsExtraP().unlink)(innerPackageFile), (0, _fsExtraP().unlink)(distInfoFile)]);
    packager.dispatchArtifactCreated(artifactPath, this, arch, packager.computeSafeArtifactName(artifactName, "pkg", arch));
  }

  async customizeDistributionConfiguration(distInfoFile, appPath) {
    await (0, _builderUtil().exec)("productbuild", ["--synthesize", "--component", appPath, distInfoFile], {
      cwd: this.outDir
    });
    const options = this.options;
    let distInfo = await (0, _fsExtraP().readFile)(distInfoFile, "utf-8");
    const insertIndex = distInfo.lastIndexOf("</installer-gui-script>");
    distInfo = distInfo.substring(0, insertIndex) + `    <domains enable_anywhere="${options.allowAnywhere}" enable_currentUserHome="${options.allowCurrentUserHome}" enable_localSystem="${options.allowRootDirectory}" />\n` + distInfo.substring(insertIndex);
    const license = await (0, _license().getNotLocalizedLicenseFiles)(options.license, this.packager);

    if (license != null) {
      distInfo = distInfo.substring(0, insertIndex) + `    <license file="${license}"/>\n` + distInfo.substring(insertIndex);
    }

    (0, _builderUtil().debug)(distInfo);
    await (0, _fsExtraP().writeFile)(distInfoFile, distInfo);
  }

  async buildComponentPackage(appPath, outFile) {
    const options = this.options;
    const args = ["--component", appPath];
    (0, _builderUtil().use)(this.options.installLocation || "/Applications", it => args.push("--install-location", it));

    if (options.scripts != null) {
      args.push("--scripts", path.resolve(this.packager.info.buildResourcesDir, options.scripts));
    } else if (options.scripts !== null) {
      const dir = path.join(this.packager.info.buildResourcesDir, "pkg-scripts");
      const stat = await (0, _fs().statOrNull)(dir);

      if (stat != null && stat.isDirectory()) {
        args.push("--scripts", dir);
      }
    }

    args.push(outFile);
    await (0, _builderUtil().exec)("pkgbuild", args);
  }

}

exports.PkgTarget = PkgTarget;

function prepareProductBuildArgs(identity, keychain) {
  const args = [];

  if (identity != null) {
    args.push("--sign", identity.hash);

    if (keychain != null) {
      args.push("--keychain", keychain);
    }
  }

  return args;
}
